pkg/endpoint/redirect_test.go

```diff
@@ -0,0 +1,256 @@
+	repo *policy.Repository
+}
+
+// GetPolicyRepository returns the policy repository of the owner.
+func (d *DummyOwner) GetPolicyRepository() *policy.Repository {
+	return d.repo
+}
+
+// QueueEndpointBuild does nothing.
+func (d *DummyOwner) QueueEndpointBuild(ctx context.Context, epID uint64) (func(), error) {
+	return nil, nil
+}
+
+// GetCompilationLock does nothing.
+func (d *DummyOwner) GetCompilationLock() *lock.RWMutex {
+	return nil
+}
+
+// SendNotification does nothing.
+func (d *DummyOwner) SendNotification(typ monitorAPI.AgentNotification, text string) error {
+	return nil
+}
+
+// Datapath returns a nil datapath.
+func (d *DummyOwner) Datapath() datapath.Datapath {
+	return nil
+}
+
+// GetNodeSuffix does nothing.
+func (d *DummyOwner) GetNodeSuffix() string {
+	return ""
+}
+
+// UpdateIdentities does nothing.
+func (d *DummyOwner) UpdateIdentities(added, deleted cache.IdentityCache) {}
+
+func (s *RedirectSuite) TestAddVisibilityRedirects(c *check.C) {
+	// Setup dependencies for endpoint.
+	kvstore.SetupDummy("etcd")
+	defer kvstore.Close()
+
+	identity.InitWellKnownIdentities()
+	idAllocatorOwner := &DummyIdentityAllocatorOwner{}
+
+	<-cache.InitIdentityAllocator(idAllocatorOwner, nil, nil)
+	defer cache.Close()
+
+	do := &DummyOwner{
+		repo: policy.NewPolicyRepository(),
+	}
+	identitymanager.Subscribe(do.repo)
+
+	lblQA := labels.ParseLabel("QA")
+	lblBar := labels.ParseLabel("bar")
+
+	httpPort := uint16(19001)
+	dnsPort := uint16(19002)
+	kafkaPort := uint16(19003)
+
+	rsp := &RedirectSuiteProxy{
+		parserProxyPortMap: map[policy.L7ParserType]uint16{
+			policy.ParserTypeHTTP:  httpPort,
+			policy.ParserTypeDNS:   dnsPort,
+			policy.ParserTypeKafka: kafkaPort,
+		},
+		redirectPortUserMap: make(map[uint16][]string),
+	}
+
+	ep := NewEndpointWithState(do, rsp, 12345, StateRegenerating)
+
+	qaBarLbls := labels.Labels{lblBar.Key: lblBar, lblQA.Key: lblQA}
+	epIdentity, _, err := cache.AllocateIdentity(context.Background(), idAllocatorOwner, qaBarLbls)
+	c.Assert(err, check.IsNil)
+	ep.SetIdentity(epIdentity, true)
+
+	firstAnno := map[string]string{
+		annotation.ProxyVisibility: "<Ingress/80/TCP/HTTP>",
+	}
+	ep.UpdateVisibilityPolicy(firstAnno)
+	err = ep.regeneratePolicy()
+	c.Assert(err, check.IsNil)
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	cmp := completion.NewWaitGroup(ctx)
+
+	_, err, _, _ = ep.addNewRedirects(ep.desiredPolicy.L4Policy, cmp)
+	c.Assert(err, check.IsNil)
+	v, ok := ep.desiredPolicy.PolicyMapState[policy.Key{
+		Identity:         0,
+		DestPort:         uint16(80),
+		Nexthdr:          uint8(u8proto.TCP),
+		TrafficDirection: trafficdirection.Ingress.Uint8(),
+	}]
+	c.Assert(ok, check.Equals, true)
+	c.Assert(v.ProxyPort, check.Equals, httpPort)
+
+	secondAnno := map[string]string{
+		annotation.ProxyVisibility: "<Ingress/80/TCP/Kafka>",
+	}
+
+	ep.UpdateVisibilityPolicy(secondAnno)
+	d, err, _, _ := ep.addNewRedirects(ep.desiredPolicy.L4Policy, cmp)
```

Comment:
Doesn't `secondAnno` conflict with `firstAnno`? Should it return an error?

`firstAnno`: `"<Ingress/80/TCP/HTTP>"`
`secondAnno`: `"<Ingress/80/TCP/Kafka>"`

Reply:
The value provided to `UpdateVisibilityPolicy` describes the desired state; this means that if we call `UpdateVisibilityPolicy(firstAnno)`, and `UpdateVisibilityPolicy(secondAnno)`, the redirect from firstAnno will be removed, as secondAnno describes the desired state. Conflicts can only occur if in the entire annotation string there is a conflict, e.g. `<Ingress/80/TCP/HTTP>,<Ingress/80/TCP/Kafka>`

Reply:
I read this the wrong way, makes sense. Thanks for explaining.